extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/Liquibase.java

```diff
@@ -0,0 +1,932 @@
+        return executeFunction(t -> t.calculateCheckSum(changeSetIdentifier));
+    }
+
+    /**
+     * Implementation of the {@link liquibase.Liquibase#calculateCheckSum(String, String, String)} ()} method.
+     * 
+     * @param filename the file name
+     * @param id the change if
+     * @param author the author
+     * @return the check sum of the change
+     * @throws LiquibaseException if the method fails.
+     */
+    public CheckSum calculateCheckSum(String filename, String id, String author) throws LiquibaseException {
+        return executeFunction(t -> t.calculateCheckSum(filename, id, author));
+    }
+
+    /**
+     * Implementation of the {@link liquibase.Liquibase#diff(Database, Database, CompareControl)} ()} method.
+     * 
+     * @param referenceDatabase the reference database
+     * @param targetDatabase the target database
+     * @param compareControl the compare control
+     * @return the diff result
+     * @throws LiquibaseException if the method fails
+     */
+    public DiffResult diff(Database referenceDatabase, Database targetDatabase, CompareControl compareControl)
+            throws LiquibaseException {
+        return executeFunction(t -> t.diff(referenceDatabase, targetDatabase, compareControl));
+    }
+
+    /**
+     * Implementation of the {@link liquibase.Liquibase#validate()} method.
+     * 
+     * @throws LiquibaseException if the method fails.
+     */
+    void validate() throws LiquibaseException {
+        executeConsumer(liquibase.Liquibase::validate);
+    }
+
+    /**
+     * Creates the default labels base on the configuration
+     * 
+     * @return the label expression
+     */
+    private LabelExpression createLabels() {
+        return new LabelExpression(config.labels);
+    }
+
+    /**
+     * Creates the default contexts base on the configuration
+     * 
+     * @return the contexts
+     */
+    private Contexts createContexts() {
+        return new Contexts(config.contexts);
+    }
+
+    /**
+     * Creates the liquibase
+     * 
+     * @param con the database connection
+     * @return the corresponding liquibase instance
+     * @throws DatabaseException if the method fails
+     */
+    private liquibase.Liquibase createLiquibase(Connection con) throws DatabaseException {
+        Database database = DatabaseFactory.getInstance().findCorrectDatabaseImplementation(new JdbcConnection(con));
+        database.setDatabaseChangeLogLockTableName(config.databaseChangeLogLockTableName);
+        database.setDatabaseChangeLogTableName(config.databaseChangeLogTableName);
+        config.currentDateTimeFunction.ifPresent(database::setCurrentDateTimeFunction);
+        config.liquibaseCatalogName.ifPresent(database::setLiquibaseCatalogName);
+        config.liquibaseSchemaName.ifPresent(database::setLiquibaseSchemaName);
+        config.liquibaseTablespaceName.ifPresent(database::setLiquibaseTablespaceName);
```

Comment:
Is it a good idea to create a Liquibase instance for each operation?

I'm not familiar with Liquibase but it doesn't seem right to me. I'm not sure I see another solution though as I suppose we need to close it somehow but I thought I should raise the issue.

Reply:
@gsmet I am not sure but I found another option which could be the one

I am using the data-source driver to create a liquibase database and liquibase instance is created only once. 

Then we have in each Liquibase bean method this logic:
Open JDBC connection
Execute liquibase.Liquibase method
Close JDBC connection

https://github.com/andrejpetras/quarkus/blob/de19e6e3de27ce9ede78069addce066dfb8c483e/extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/Liquibase.java#L65

Like hibernate-orm or quartz extension I have guess method for the liquibase database:
https://github.com/andrejpetras/quarkus/blob/de19e6e3de27ce9ede78069addce066dfb8c483e/extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/Liquibase.java#L105

I am not sure if it is possible to move this part in the LiquibaseProcessor but I found this:
https://github.com/quarkusio/quarkus/blob/f8ccf737a65461b4c123e961986a3d4c35adefd6/extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java#L113

It looks like DataSourceDriverBuildItem currently supports only a default datasource

Commit: 
https://github.com/andrejpetras/quarkus/commit/de19e6e3de27ce9ede78069addce066dfb8c483e

Reply:
@gsmet could you please check the solution in this commit? https://github.com/andrejpetras/quarkus/commit/de19e6e3de27ce9ede78069addce066dfb8c483e
Looks like this is the last open point in this PR

Reply:
Only the JDBC connection is created for each method execution

Reply:
Let's not mark conversation as resolved if we are discussing it :). Sorry, I was away for a bit, will have a look at your commit.

Reply:
I'm not convinced it's a good idea to hardcode the databases here. We want to get rid of it elsewhere so I don't think generalizing it is a good idea.

What I was wondering is if we should have something like:

```java
@Inject
LiquibaseHandler liquibaseHandler;

public void myMethod() {
    try (Liquibase liquibase = liquibaseHandler.open()) {
         // do your Liquibase operations here
    }
}
```

The `LiquibaseHandler` name might not be ideal but it's the pattern I'm interested in. We would use a proper try with resource to handle the close (providing `Liquibase` is closeable, I'm not sure?).

That way the connection would be short lived but we could still reuse the same connection for several operations and we wouldn't have to wrap the `Liquibase` class.

WDYT? /cc @gastaldi

Reply:
BTW, it's just an idea, I haven't checked it could work so maybe check if it could be doable but let's discuss it with you and @gastaldi before doing too much work.

Reply:
AFAIR there isn't a concept of "closing" a Liquibase object in the Liquibase API, but I may be wrong (it's been a while since I used it). I think the connection is kept open until it `DatabaseConnection.close` is explicitly closed

Reply:
@gsmet sorry for the "resolved conversation".

Please take a look at the current implementation in PR
extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/Liquibase.java

Reply:
But yeah, I can see why a try-with-resources block may be useful here

Reply:
> AFAIR there isn't a concept of "closing" a Liquibase object in the Liquibase API, but I may be wrong (it's been a while since I used it). I think the connection is kept open until it `DatabaseConnection.close` is explicitly closed

Yes, this: ```Liquibase.getDatabase().close();```

Reply:
>The LiquibaseHandler name might not be ideal but it's the pattern I'm interested in. We would use a proper try with resource to handle the close (providing Liquibase is closeable, I'm not sure?).

The Liquibase object doesn't implement Closeable.

Also some of the methods like dropAll are final
```
public final void dropAll() throws DatabaseException {
        this.dropAll(new CatalogAndSchema(this.getDatabase().getDefaultCatalogName(), this.getDatabase().getDefaultSchemaName()));
    }
```

Reply:
@gsmet the ```LiquibaseHandler``` makes sense to me.

This should be possible:
```
package org.quarkus.liquibase;
public class Liquibase implements Closeable {
}
```

```
@Inject
LiquibaseHandler liquibaseHandler;

public void myMethod() {
    try (org.quarkus.liquibase.Liquibase liquibase = liquibaseHandler.open()) {
         // do your Liquibase operations here
    }
}
```

Reply:
Another possibility is to leave the closing operations away from the user code. Say you have something like:

```java
public interface LiquibaseConsumer {
    void consume(Liquibase liquibase);
}
```

And in the extension you could iterate through all the LiquibaseConsumer implementations and close the connection afterwards

Reply:
Although I have seen in practice that users don't really use the Liquibase object, they just put the changelog.xml files in a specific folder and Liquibase automatically uses them to perform the necessary DB operations. (Through a [LiquibaseServletListener](https://www.liquibase.org/javadoc/liquibase/integration/servlet/LiquibaseServletListener.html))

Reply:
mmh, what exactly would be the point where we wanted to close the connection?

Reply:
Would it be possible to remove the "final" flag from the Liquibase method by modifying the byte code?

Reply:
I can try this:
```
public class LiquibaseImpl extends liquibase.Liquibase implements Closeable { }
```

```
    try (liquibase.Liquibase liquibase = liquibaseHandler.open()) {
         // do your Liquibase operations here
    }
```

Reply:
> mmh, what exactly would be the point where we wanted to close the connection?

What would be the use case where you'll want to close the connection before all changes are performed?

Reply:
For example, the user can start the migration and then read the Liquibase information about the changes.

```
liquibase.update();
List<ChangeSetStatus> status = liquibase.getChangeSetStatuses();
```

Reply:
@gsmet @gastaldi 

Please check this:
https://github.com/andrejpetras/quarkus/commit/e061b821b4fe60084f1bccbae0acf60c75177738

```java
    @Inject
    LiquibaseFactory liquibaseFactory;

    try (LiquibaseContext liquibase = liquibaseFactory.createContext()) {
         liquibase.update();
    }
```

- no hardcode databases
- the names LiquibaseFactory and LiquibaseContext might not be ideal.

Reply:
That looks cool

Reply:
Should I merge it into PR?

Reply:
@gsmet please have a look at the solution.

Reply:
PR update for liquibase 3.8.7 version
```   @Inject
    LiquibaseFactory liquibaseFactory;

    try (Liquibase liquibase = liquibaseFactory.createLiquibase()) {
         liquibase.update(liquibaseFactory.createContexts(),liquibaseFactory.createLabels());
    }```

Reply:
The liquibase version 3.8.7 will be release in the next few weeks
Liquibase PR: https://github.com/liquibase/liquibase/pull/987